### PR TITLE
API cleanup & performance passes

### PR DESCRIPTION
**Pull request by earhart**

### CODE REVIEW

File: src/ngraph/runtime/plaidml/plaidml_pass_lower_convolutions.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/pattern/matcher.hpp"
+#include "ngraph/pattern/op/label.hpp"
+#include "ngraph/runtime/plaidml/plaidml_ops_convolution.hpp"
+#include "ngraph/runtime/plaidml/plaidml_pass_lower_convolutions.hpp"
+
+ngraph::runtime::plaidml::pass::LowerConvolutions::LowerConvolutions()
+{
+    auto convolution_op =
+        std::make_shared<pattern::op::Label>(element::i8, Shape{}, [](std::shared_ptr<Node> node) {
```
Comment (Krovatkin) [inline src/ngraph/runtime/plaidml/plaidml_pass_lower_convolutions.cpp]: I still wonder if `Reshape->Any->Reshape` might be a better choice?

File: src/ngraph/runtime/plaidml/plaidml_pass_reshape_elision.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        Shape{},
+        [](std::shared_ptr<Node> node) {
+            op::Reshape* reshape = dynamic_cast<op::Reshape*>(node.get());
+            if (!reshape)
+            {
+                return false;
+            }
+
+            // Validate that this isn't a reordering-reshape.
+            for (std::size_t idx = 0; idx < reshape->get_input_order().size(); ++idx)
```
Comment (Krovatkin) [inline src/ngraph/runtime/plaidml/plaidml_pass_reshape_elision.cpp]: `get_is_transpose` ?

**Comment by earhart** [comment]
@Krovatkin Thanks for the extensive review -- it might take me some time to get to all these, but I'll work on it.  :-)

File: src/ngraph/runtime/plaidml/plaidml_compiler.cpp [hunk 3]
```diff
@@ -78,15 +89,44 @@ std::shared_ptr<ngraph::runtime::plaidml::CompiledFunction>
...
     // backprop
     pass_manager.register_pass<ngraph::pass::Liveness>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::ExplicitLogicals>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::ConcatElision>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::ReplicateElision>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::ReplicateCombination>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::ImplicitBroadcast>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::ReshapeElision>();
+    pass_manager.register_pass<ngraph::runtime::plaidml::pass::LowerConvolutions>();
+    if (m_config->winograd)
```
Comment (jbobba) [inline src/ngraph/runtime/plaidml/plaidml_compiler.cpp]: We are using a pass_config object to knob these passes. https://github.com/NervanaSystems/ngraph/blob/master/src/ngraph/runtime/cpu/cpu_external_function.cpp#L1098 
Feel free to use them if you find them useful.

File: src/ngraph/runtime/plaidml/plaidml_pass_concat_elision.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+#include "ngraph/pattern/op/any.hpp"
+#include "ngraph/pattern/op/any_of.hpp"
+#include "ngraph/pattern/op/label.hpp"
+#include "ngraph/runtime/plaidml/plaidml_ops_replicate.hpp"
+
+ngraph::runtime::plaidml::pass::ConcatElision::ConcatElision()
+{
+    auto concat_op =
+        std::make_shared<pattern::op::Label>(element::i8, Shape{}, [](std::shared_ptr<Node> node) {
+            return dynamic_cast<ngraph::op::Concat*>(node.get()) != nullptr;
```
Comment (jbobba) [inline src/ngraph/runtime/plaidml/plaidml_pass_concat_elision.cpp]: dynamic_cast doesn't seem to work properly on Macs. You could use node->description() instead.

File: src/ngraph/runtime/plaidml/plaidml_pass_concat_elision.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        });
+
+    pattern::graph_rewrite_callback callback = [](pattern::Matcher& m) {
+        auto concat = std::dynamic_pointer_cast<ngraph::op::Concat>(m.get_match_root());
```
Comment (jbobba) [inline src/ngraph/runtime/plaidml/plaidml_pass_concat_elision.cpp]: static_pointer_cast?

File: src/ngraph/runtime/plaidml/plaidml_pass_reshape_elision.hpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// A minor pass to elide unnecessary reshapes.  A reshape is
+// considered unnecessary if its output shape is the same as its input
+// shape, modulo leading size-1 axes.
+class ngraph::runtime::plaidml::pass::ReshapeElision final : public ngraph::pass::GraphRewrite
+{
+public:
+    ReshapeElision();
+};
```
Comment (jbobba) [inline src/ngraph/runtime/plaidml/plaidml_pass_reshape_elision.hpp]: can this be moved into a device independent pass in ngraph core?

File: src/ngraph/runtime/plaidml/plaidml_pass_concat_elision.cpp (hunk 4 above)
Comment (earhart) [reply src/ngraph/runtime/plaidml/plaidml_pass_concat_elision.cpp]: I did all of the testing for this on a Mac, and `dynamic_cast<>` seems like something the compiler team would make really sure to get right before release... I've been following the conversation on Slack; I haven't done enough experimenting to be able to say that anyone's wrong, but... this seems to be working fine for me here.

File: src/ngraph/runtime/plaidml/plaidml_compiler.cpp (hunk 3 above)
Comment (earhart) [reply src/ngraph/runtime/plaidml/plaidml_compiler.cpp]: More environment variables?  :-). Okay, sounds good.

**Pull request merged**